ERCS/erc-7943.md

```diff
@@ -0,0 +1,501 @@
+
+    /// @notice Takes tokens from one address and transfers them to another.
+    /// @dev Requires specific authorization. Used for regulatory compliance or recovery scenarios.
+    /// @param from The address from which `amount` is taken.
+    /// @param to The address that receives `amount`.
+    /// @param tokenId The ID of the token being transferred.
+    /// @param amount The amount to force transfer.
+    function forceTransfer(address from, address to, uint256 tokenId, uint256 amount) external;
+
+    /// @notice Changes the frozen status of `amount` of `tokenId` tokens belonging to an `user`.
+    /// This overwrites the current value, similar to an `approve` function.
+    /// @dev Requires specific authorization. Frozen tokens cannot be transferred by the user.
+    /// @param user The address of the user whose tokens are to be frozen/unfrozen.
+    /// @param tokenId The ID of the token to freeze/unfreeze.
+    /// @param amount The amount of tokens to freeze/unfreeze. 
+    function setFrozen(address user, uint256 tokenId, uint256 amount) external;
+
+    /// @notice Checks the frozen status/amount of a specific `tokenId`.
+    /// @param user The address of the user.
+    /// @param tokenId The ID of the token.
+    /// @return amount The amount of `tokenId` tokens currently frozen for `user`.
+    function getFrozen(address user, uint256 tokenId) external view returns (uint256 amount);
+ 
+    /// @notice Checks if a transfer is currently possible according to token rules. It enforces validations on the frozen tokens.
+    /// @dev This may involve checks like allowlists, blocklists, transfer limits and other policy-defined restrictions.
+    /// @param from The address sending tokens.
+    /// @param to The address receiving tokens. 
+    /// @param tokenId The ID of the token being transferred.
+    /// @param amount The amount being transferred.
+    /// @return allowed True if the transfer is allowed, false otherwise.
+    function isTransferAllowed(address from, address to, uint256 tokenId, uint256 amount) external view returns (bool allowed);
+
+    /// @notice Checks if a specific user is allowed to interact according to token rules.
+    /// @dev This is often used for allowlist/KYC/KYB/AML checks.
+    /// @param user The address to check.
+    /// @return allowed True if the user is allowed, false otherwise.
+    function isUserAllowed(address user) external view returns (bool allowed);
+}
+```
+
+### `isUserAllowed`, `isTransferAllowed` and `getFrozen`
+
+These provide views into the implementing contract's compliance, transfer policy logic and freezing status. These functions:
+
+- MUST NOT revert. 
+- MUST NOT change the storage of the contract. 
+- MAY depend on context (e.g., current timestamp, block number).
+- The `isTransferAllowed` MUST validate that the `amount` being transferred doesn't exceed the unfrozen amount (which is the difference between the current balance and the frozen balance).
```

Comment:
I would also make explicit whether we expect `isTransferAllowed(from, to, tokenId, amount)` to check `isUserAllowed(to)`.

That is, whether an integrator can only rely on `isTransferAllowed` to know if a given transfer will succeed or must also call `isUserAllowed`.

I'd go with the first.